File: src/main/java/work/lclpnet/ap2/impl/game/TeamEliminationGameInstance.java
package work.lclpnet.ap2.impl.game;

import net.fabricmc.fabric.api.networking.v1.PlayerLookup;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayerEntity;
import work.lclpnet.ap2.api.base.Participants;
import work.lclpnet.ap2.api.game.MiniGameHandle;
import work.lclpnet.ap2.api.game.data.DataContainer;
import work.lclpnet.ap2.api.game.team.Team;
import work.lclpnet.ap2.api.game.team.TeamKey;
import work.lclpnet.ap2.api.game.team.TeamManager;
import work.lclpnet.ap2.impl.game.data.EliminationDataContainer;
import work.lclpnet.ap2.impl.game.data.type.TeamRef;
import work.lclpnet.ap2.impl.util.DeathMessages;
import work.lclpnet.kibu.hook.entity.EntityHealthCallback;
import work.lclpnet.kibu.plugin.hook.HookRegistrar;
import work.lclpnet.kibu.translate.TranslationService;
import work.lclpnet.lobby.game.api.WorldFacade;

import java.util.HashSet;
import java.util.Set;

import static net.minecraft.util.Formatting.GRAY;

public abstract class TeamEliminationGameInstance extends DefaultTeamGameInstance {

    private final EliminationDataContainer<Team, TeamRef> data = new EliminationDataContainer<>(this::createReference);

    public TeamEliminationGameInstance(MiniGameHandle gameHandle) {
        super(gameHandle);
    }

    /**
     * Instantly makes players who would have died spectators and reset them.
     */
    protected final void useSmoothDeath() {
        HookRegistrar hooks = gameHandle.getHookRegistrar();

        hooks.registerHook(EntityHealthCallback.HOOK, (entity, health) -> {
            if (!(entity instanceof ServerPlayerEntity player) || health > 0) return false;

            eliminate(player);

            return true;
        });
    }

    protected void eliminate(ServerPlayerEntity player) {
        Participants participants = gameHandle.getParticipants();

        if (participants.isParticipating(player)) {
            DeathMessages deathMessages = gameHandle.getDeathMessages();
            MinecraftServer server = gameHandle.getServer();

            deathMessages.eliminated(player).sendTo(PlayerLookup.all(server));

            participants.remove(player);
        }

        resetPlayer(player);
    }

    private void resetPlayer(ServerPlayerEntity player) {
        WorldFacade worldFacade = gameHandle.getWorldFacade();
        PlayerUtil playerUtil = gameHandle.getPlayerUtil();

        playerUtil.resetPlayer(player);
        worldFacade.teleport(player);
    }

    protected void eliminate(Team team) {
        TeamManager teamManager = getTeamManager();

        if (teamManager.isParticipating(team)) {
            DeathMessages deathMessages = gameHandle.getDeathMessages();
            MinecraftServer server = gameHandle.getServer();

            deathMessages.eliminated(team).sendTo(PlayerLookup.all(server));

            teamManager.setTeamEliminated(team);
        }

        Participants participants = gameHandle.getParticipants();

        for (ServerPlayerEntity player : team.getPlayers()) {
            participants.remove(player);
            resetPlayer(player);
        }
    }

    protected void eliminateAll(Iterable<? extends Team> teams) {
        TeamManager teamManager = getTeamManager();
        TranslationService translations = gameHandle.getTranslations();

        Set<Team> toEliminate = new HashSet<>();

        for (Team team : teams) {
            if (!teamManager.isParticipating(team)) continue;

            TeamKey key = team.getKey();
            var displayName = translations.translateText(key.getTranslationKey()).formatted(key.colorFormat());

            translations.translateText("ap2.game.team_eliminated", displayName)
                    .formatted(GRAY)
                    .sendTo(PlayerLookup.all(gameHandle.getServer()));

            toEliminate.add(team);
        }

        // mark all teams as eliminated at the same moment
        data.allEliminated(toEliminate);

        toEliminate.forEach(teamManager::setTeamEliminated);

        Participants participants = gameHandle.getParticipants();

        // deliberately use teams instead of toEliminate, to make sure there are no participating members anymore
        for (Team team : teams) {
            for (ServerPlayerEntity player : team.getPlayers()) {
                participants.remove(player);
                resetPlayer(player);
            }
        }
    }

    @Override
    public void teamEliminated(Team team) {
        // make sure the team is tracked as eliminated
        data.eliminated(team);

        super.teamEliminated(team);
    }

    @Override
    public DataContainer<Team, TeamRef> getData() {
        return data;
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/base/TeamEliminationGameInstance.kt
package work.lclpnet.ap2.game.base

import net.minecraft.ChatFormatting
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import work.lclpnet.ap2.api.game.team.Team
import work.lclpnet.ap2.api.game.team.TeamManager
import work.lclpnet.ap2.core.hook.PlayerEliminatedCallback
import work.lclpnet.ap2.ext.allPlayers
import work.lclpnet.ap2.ext.hooks
import work.lclpnet.ap2.ext.isParticipating
import work.lclpnet.ap2.ext.translate
import work.lclpnet.ap2.game.MiniGameHandle
import work.lclpnet.ap2.game.data.EliminationDataContainer
import work.lclpnet.game.map.GameMap
import work.lclpnet.kibu.hook.entity.EntityHealthCallback
import work.lclpnet.kibu.translate.text.TranslatedText

abstract class TeamEliminationGameInstance(
    gameHandle: MiniGameHandle,
    world: ServerLevel,
    map: GameMap,
    teamManager: TeamManager
) : TeamGameInstance(gameHandle, world, map, teamManager) {

    override val data = EliminationDataContainer { team: Team ->
        createReference(team)
    }

    /**
     * Instantly makes players who would have died spectators and reset them.
     */
    protected fun useSmoothDeath() {
        EntityHealthCallback.HOOK.registerWith(hooks) { entity, health ->
            if (entity !is ServerPlayer || health > 0) return@registerWith false
            eliminate(entity)
            true
        }
    }

    protected fun eliminate(player: ServerPlayer) {
        if (isParticipating(player)) {
            gameHandle.deathMessages.eliminated(player).sendTo(allPlayers())

            gameHandle.participants.remove(player)
            onEliminated(player)
        }

        resetPlayer(player)
    }

    private fun resetPlayer(player: ServerPlayer) {
        gameHandle.playerUtil.resetPlayer(player)
        gameHandle.worldFacade.teleport(player)
    }

    protected fun eliminate(team: Team) {
        // eliminate all remaining team players first
        for (player in team.getPlayers()) {
            gameHandle.participants.remove(player)
            onEliminated(player)
            resetPlayer(player)
        }

        // now actually eliminate the team
        if (!teamManager.isParticipating(team)) return

        gameHandle.deathMessages.eliminated(team).sendTo(allPlayers())

        teamManager.setTeamEliminated(team)
    }

    protected fun eliminateAll(teams: Iterable<Team>, detail: TranslatedText? = null) {
        val toEliminate = mutableSetOf<Team>()

        for (team in teams) {
            if (!teamManager.isParticipating(team)) continue

            val key = team.key()
            val displayName = key.getDisplayName(gameHandle.translations)

            translate("ap2.game.team_eliminated", displayName)
                .formatted(ChatFormatting.GRAY)
                .sendTo(allPlayers())

            toEliminate.add(team)
        }

        // mark all teams as eliminated at the same moment
        data.addAll(toEliminate, detail)

        // deliberately use teams instead of toEliminate, to make sure there are no participating members anymore
        for (team in teams) {
            for (player in team.getPlayers()) {
                gameHandle.participants.remove(player)
                onEliminated(player)
                resetPlayer(player)
            }
        }

        // finally actually set the teams to eliminated, after the all team participants are eliminated
        for (team in toEliminate) {
            teamManager.setTeamEliminated(team)
        }
    }

    override fun teamEliminated(team: Team) {
        // make sure the team is tracked as eliminated
        data.add(team)

        super.teamEliminated(team)
    }

    protected fun onEliminated(player: ServerPlayer) {
        PlayerEliminatedCallback.HOOK.invoker().onEliminated(player)
    }
}